streaming/java/streaming-state/src/main/java/org/ray/streaming/state/strategy/CPStateStrategy.java

```diff
@@ -0,0 +1,143 @@
+package org.ray.streaming.state.strategy;
+
+import com.google.common.primitives.Longs;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import org.ray.streaming.state.StateException;
+import org.ray.streaming.state.StorageRecord;
+import org.ray.streaming.state.keystate.state.impl.AbstractState;
+import org.ray.streaming.state.store.IKVStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class define the checkpoint store strategy, which saves two-version data once.
+ */
+public class CPStateStrategy<V> extends AbstractStateStrategy<V> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(CPStateStrategy.class);
+
+  public CPStateStrategy(IKVStore<String, Map<Long, byte[]>> backStore) {
+    super(backStore);
+  }
+
+  @Override
+  public void finish(long batchId) {
+    LOGGER.info("do finish batchId:{}", batchId);
+    Map<String, byte[]> cpStore = new HashMap<>();
+    for (Entry<String, StorageRecord<V>> entry : frontStore.entrySet()) {
+      String key = entry.getKey();
+      StorageRecord<V> value = entry.getValue();
+      cpStore.put(key, toByte(value));
+    }
+    middleStore.put(batchId, cpStore);
+    frontStore.clear();
+  }
+
+  @Override
+  public Object commit(long batchId, Object state) {
+    try {
+      LOGGER.info("do commit batchId:{}", batchId);
+      Map<String, byte[]> cpStore = middleStore.get(batchId);
+      if (cpStore == null) {
+        throw new StateException("why cp store is null");
+      }
+      for (Entry<String, byte[]> entry : cpStore.entrySet()) {
+        String key = entry.getKey();
+        byte[] value = entry.getValue();
+
+        // 2 is specific key in kv store and indicates that new value
```

Comment:
Add more comments for multi-version state.

Reply:
ok